Added Network Table / Node refactoring

turcato-niccolo opened the pull request.

Commit:
Updates to Peer, Message, Resource and implementing classes.

Peer, Message and Resource are now abstract classes overriding equals and
CompareTo. Fixed SMSPeer, SMSMessage and StringResource according to such changes.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

ricdezen committed:
Merge pull request #23 from RicDeZen/feature/networkmanager

Feature/networkmanager

Commit:
Merge branch 'development' of https://github.com/RicDeZen/SmsHandler into feature/abs-peer-message-res

Commit:
Fixed specs for sendUrgentMessage()

Fixed specs for sendUrgentMessage() in SMSManager
Fixed specs for sendMessage() in SMSHandler

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

ricdezen committed:
Merge pull request #24 from RicDeZen/feature/abs-peer-message-res

Peer, Message, Resource as abstract classes

Commit:
Fix for missing line in SMSManager.onContextDestroyed()

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
PendingIntents in SMSHandler are now unique

A request code is generated using current system time. It is not a perfect solution but it minimizes error chance.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

ricdezen committed:
Merge pull request #26 from RicDeZen/hotfix/development-pendingIntents

PendingIntents in SMSHandler are now unique

Commit:
Updated specs for SMSHandler.onContextDestroyed

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Fix for SMSManager.onContextDestroyed(), getInstance() and isOwner()

The class now always uses the application context, to allow certainty
regarding who the owner of the instance is (always the parent application)

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Implementation of class for operations on database containing unread sms.

- SMSDatabaseManager, singleton class handling the operations on said database.
- Renamed SMSDao,SMSDatabase,SMSEntity to match the rest of the package.
- Made three said classes package-private.
- Modified SMSHandler, SMSManager and SMSReceiver to handle fetching of unread messages through the new class.
- Removed method from SmsUtils

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

ricdezen committed:
Merge pull request #28 from RicDeZen/feature/smshandler-smsdatabase

Implementation of class for operations on database containing unread sms.

Commit:
Removed SMSManager.isOwner()

The method is not necessary anymore because of the new ownership mechanism which uses always the application context.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Started implementing reflection to start an activity from saved value.

Created MEFileManager class as an object pool to read and write from files with atomic operations through synchronized methods.
Modified SMSReceiver to start an Activity instead of firing a Broadcast.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Modified implementation of reflection.

Moved MEFileManager to another module for later use.
Created PreferenceManager as a class meant to make storing Preferences easier.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Completed implementation of reflection.

Moved PreferenceManager in module together with MEFileManager for better generic implementation of both.
Reflection has been tested to work empirically.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Changes to SMSManager and SMSHandler methods to set the value of the activity.

Said methods now allow overloading on both Class and Activity and check if the Class object represents a class extending Activity.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

turcato-niccolo committed:
Merge pull request #30 from RicDeZen/feature/smshandler-reflection

Reflection taking place of Broadcast.

Commit:
Fixed a specific in SMSManager

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Fixes to NetworkManager

- Added InvalidMsgSyntaxException.
- Divided sending action into: composing the body, sending to a peer, broadcasting.
- Turned Actions constants into int values.
- Turned subsequent ifs into a switch.
- Moved action interpretation to separate method.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Minor rearrangement to SMSReceiver

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Minor fixes to SMSReceiver

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

ricdezen committed:
Extensive validity check for SMSPeer and SMSMessage (#34)

* Added enums to detect anomalies in SMSPeer and SMSMessage.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

* Added Tests for SMSPeer and SMSMessage.

Fixed mistakes located through tests.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

* Fixed code review changes

- Missing @throws in SMSPeer constructor
- Missing @throws in SMSMessage constructor
- Missing peer.isValid() check in SMSMessage constructor

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Minor specs fixes

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

ricdezen committed:
Merge pull request #35 from RicDeZen/hotfix/smshandler-smsreceiver

Fixed methods and specs in SMSReceiver

Commit:
Working Resource synchronization for already connected Peers.

- Pulled from feature/smshandler
- Modified Interface and Demo to show Resources

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Fixed typo

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

ricdezen committed:
Merge pull request #38 from RicDeZen/hotfix/networkmanager-constants

NetworkManager Resources and Int constants

turcato-niccolo committed:
Fixed equals method in Node, PeerNode, ResourceNode

turcato-niccolo committed:
Feature/networkmanager turcato (#39)

Added isGlobalPhoneNumber in SMSPeer to replace PhoneNumberUtils.isGlobalPhoneNumber()

Added removeKeysFromMessageBody, for both Strings and SMSMessages in SmsUtils

Added equals in SMSMessage

Added tracking of deleted resources and Peers in NetworkDictionary, for a correct export to backup Database

Fixed some old tests

Fixed equals method in SMSMessage, after Giorgia Bortoletti's review

Commit:
Added method to send the existing resources to a new Peer.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Merge remote-tracking branch 'origin/feature/networkmanager' into feature/networkmanager

turcato-niccolo committed:
Merge branch 'feature/networkmanager' into feature/distributedNetwork

# Conflicts:
#	networkmanager/src/main/java/com/dezen/riccardo/networkmanager/NetworkManager.java

turcato-niccolo committed:
-Changed structure for Node classes, now Node is an abstract class, ist realization is DistributedNetworkNode, which has replaced previous Node class.

-Now KBucket works with abstract Node objects

-Added abstract class for Routing Table (Kademlia like p2p network) which uses abstract structure Bucket<Node>.

-PeerNode now extends Peer<DistributedNetworkNode>

-ResourceNode has been modified to use DistributedNetworkNode as address, in accordance with extended abstract class Resource<K, V>

bortoletti-giorgia commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkNode.java:
```diff
@@ -0,0 +1,167 @@
+     * @return the bitSet containing the hash of the peer's address, bitSet's length is a multiple of 64
+     * @throws IllegalArgumentException if peer isn't valid or the numBit isn't multiple of 64
+     */
+    public static BitSet hash(@NonNull String toHash, int numBits) {
```
In this method you can use some encryption algorithm as SHA-1 for a key of 160 bits or MD5 for a key of 128 bit.
Like this:
```
try{
            MessageDigest md = MessageDigest.getInstance("SHA-1");
            md.update(toHash.getBytes());
            byte[] digest = md.digest();

            id = fromByteArray(digest);
}catch(NoSuchAlgorithmException e){}
```

---and---
```
/**
     * Returns a bitset containing the values in bytes.
     * @param bytes to convert in BitSet object
     */
    public static BitSet fromByteArray(byte[] bytes) {
      BitSet bitSet = new BitSet();
      for (int i = 0; i < bytes.length * 8; i++) {
        if ((bytes[bytes.length - i / 8 - 1] & (1 << (i % 8))) > 0) {
        	bitSet.set(i);
        }
      }
      return bitSet;
    }
```

turcato-niccolo replied:
This is much better! Thanks!

turcato-niccolo committed:
Merge branch 'feature/distributedNetwork' into feature/distributedNetwork-RoutingTable

bortoletti-giorgia commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/NetworkDictionaryWithOwnership.java:
```diff
@@ -20,15 +20,15 @@
 
     /**
      * Constructor, initializes the object and defines the backup database
+     *
```
In this class I would write Map<Peer, Node[]>.

bortoletti-giorgia commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/RoutingTable.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * @return true if the node has been added, false otherwise
+     */
+    public abstract boolean Add(Node node);
```
It is a minimal correction but each method should start with a lowercase letter.

bortoletti-giorgia commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KBucket.java:
```diff
@@ -1,56 +1,64 @@
+
+public class KBucket extends Bucket<Node> {
+
     private int dimension;
```
"dimension" is present also in Bucket<Node> so it isn't necessary to insert it and just add a getDimension() on Bucket<Node> and on the constructor of KBucket invoke "super(dim)".

RicDeZen commented:
Sorry, I've been working on this for a while and I can't seem to find a good reason to define our bucket ourselves, we have no iterator, no multiple adding and removal, we simply created a class that does less than what an ArrayList would already do. Also the Nodes are very complicated, what do we use in the network? PeerNode and ResourceNode or ReplicatedNetworkNode?

RicDeZen commented:
Actually I'll jump straight to adding requests and such, I can't expect to overwrite everything everyone does

turcato-niccolo commented:
> Sorry, I've been working on this for a while and I can't seem to find a good reason to define our bucket ourselves, we have no iterator, no multiple adding and removal, we simply created a class that does less than what an ArrayList would already do. 

Doing more isn't the point, for example in the routing table we need containers that do just a set of operations, for this, I've written the abstract Bucket. KBucket is a realization of that, and I see it's not complete of course but it's not useless. If we need it to be iterable, we will implement it.

> Also the Nodes are very complicated, what do we use in the network? PeerNode and ResourceNode or ReplicatedNetworkNode?

Nodes are a little complicated but if you want to manage BitSets you need something like this, as Bitsets are not of fixed lenghts, and we shoudn't accept any kind of Bitsets, as you wouldn't know which address space they belong. It's a tradeoff for having a handy binary key.
Moreover if you have seen, with DistributedNetworkNode you can calculate distances between keys, which is handy if you want to sort them.
DistributedNetworkNode is what we would use as Keys in the distributed network, both PeerNode and ResourceNode return an istance of DistributedNetworkNode with method getAddress.
It was the simplest way, that I've thought of, to have a generalized Ojbect for the addresses/keys of the distributed network and two objects that extends the abstract classes Peer<T> and Resource<K,V>. So that we could reuse objects defined on them, like maybe the dictionary used also in NetworkManager.

> I'll leave the Buckets as they are nd focus on the Nodes

Also, a small note: both Buckets and Nodes were added last week: mo' me lo dici?😂😂

Commit:
Fixed method names

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

turcato-niccolo committed:
Merge remote-tracking branch 'origin/feature/distributedNetwork' into feature/distributedNetwork-RoutingTable

# Conflicts:
#	distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KBucket.java

turcato-niccolo committed:
-Added binarySet, essentially a comparable BitSet

-Changed structure for Nodes, now theres a Node<T> interface, PeerNode and Resource node implement Node<BinarySet> and respectfully Peer<BinarySet>, Resource<BinarySet, String>

turcato-niccolo committed:
Merge remote-tracking branch 'origin/feature/distributedNetwork-RoutingTable' into feature/distributedNetwork-RoutingTable

# Conflicts:
#	distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KBucket.java
#	distributednetworkmanager/src/test/java/com/gruppo1/distributednetworkmanager/KBucketTest.java

turcato-niccolo committed:
DistributednetworkNode and NetworkDictionaryWithOwnerShip went deprecated

Added Testing for BitSetUtils, BinarySet and Nodes

Added SHA-1 Hashin in BitSetUtils
